StudyConnect.Data/Repositories/PostRepository.cs

```diff
@@ -0,0 +1,218 @@
+using Microsoft.EntityFrameworkCore;
+using StudyConnect.Core.Common;
+using StudyConnect.Core.Models;
+using StudyConnect.Core.Interfaces;
+
+namespace StudyConnect.Data.Repositories;
+
+public class PostRepository : BaseRepository, IPostRepository
+{
+    public PostRepository(StudyConnectDbContext context) : base(context)
+    {
+
+    }
+
+    public async Task<OperationResult<Guid?>> AddAsync(Guid userId, Guid categoryId, ForumPost? post)
+    {
+        if (userId == Guid.Empty)
+            return OperationResult<Guid?>.Failure("User Id is Invalid.");
+
+        if (categoryId == Guid.Empty)
+            return OperationResult<Guid?>.Failure("Category Id is Invalid.");
+
+        if (post == null)
+            return OperationResult<Guid?>.Failure("Post cannot be null.");
+
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.UserGuid == userId);
+        var category = await _context.ForumCategories.FirstOrDefaultAsync(c => c.ForumCategoryId == categoryId);
+        if (user == null || category == null)
+            return OperationResult<Guid?>.Failure("User or Category not found.");
+
+        var testTitle = await _context.ForumPosts.AnyAsync(fp => fp.Title == post.Title);
+        if (testTitle)
+            return OperationResult<Guid?>.Failure("Post Title already taken.");
+
+        try
+        {
+            var newPost = new Entities.ForumPost
+            {
+                Title = post.Title,
+                Content = post.Content,
+                ForumCategory = category,
+                User = user
+            };
+
+            await _context.ForumPosts.AddAsync(newPost);
+            await _context.SaveChangesAsync();
+
+
+            return OperationResult<Guid?>.Success(newPost.ForumPostId);
+        }
+        catch (Exception ex)
+        {
+            return OperationResult<Guid?>.Failure($"Failed to add the Post: {ex.Message}");
+        }
+    }
+
+    public async Task<OperationResult<IEnumerable<ForumPost>?>> SearchAsync(Guid? userId, string? categoryName, string? title)
+    {
+        var posts = _context.ForumPosts
+            .AsNoTracking()
+            .Include(p => p.User)
+            .Include(p => p.ForumCategory)
+            .AsQueryable();
+
+        if (userId.HasValue)
+            posts = posts.Where(p => p.User.UserGuid == userId.Value);
+
+        if (!string.IsNullOrEmpty(categoryName))
+            posts = posts.Where(p => p.ForumCategory.Name == categoryName);
+
+        if (!string.IsNullOrEmpty(title))
+            posts = posts.Where(p => EF.Functions.Like(p.Title, $"%{title}%"));
```

Comment:
Maybe want to use case insensitivity here.
```suggestion
            posts = posts.Where(p => EF.Functions.Like(p.Title.ToLower(), $"%{title}%".ToLower()));
```

Reply:
it seems that our default database collation is case-insensitive, therefore it is not needed.
Tested it out on my side.
Also .ToLower() is extreme performance heavy for database.

Reply:
You're right, my mistake it seems indeed the Database already works case insensitive